// @node-wot/binding-http/dist/codecs/tuya-codec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class HttpTuyaCodec {
    getMediaType() {
        return "application/json+tuya";
    }
    bytesToValue(bytes, schema, parameters) {
        const parsedBody = JSON.parse(bytes.toString());
        const success = parsedBody.success ?? false;
        if (!success) {
            throw new Error(parsedBody.msg ?? JSON.stringify(parsedBody));
        }
        for (const value of Object.values(parsedBody.result ?? {})) {
            if (value.code === schema["tuya:PropertyName"]) {
                return value;
            }
        }
        throw new Error("Property not found");
    }
    valueToBytes(value, schema, parameters) {
        const obj = {
            commands: [
                {
                    code: schema["tuya:PropertyName"],
                    value,
                },
            ],
        };
        const body = JSON.stringify(obj);
        return Buffer.from(body);
    }
}
exports.default = HttpTuyaCodec;
