// @node-wot/binding-http/dist/credential.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TuyaCustomBearer = exports.OAuthCredential = exports.BasicKeyCredential = exports.BearerCredential = exports.BasicCredential = exports.Credential = void 0;
const node_fetch_1 = __importStar(require("node-fetch"));
const crypto = __importStar(require("crypto"));
const queryString = __importStar(require("query-string"));
class Credential {
}
exports.Credential = Credential;
class BasicCredential extends Credential {
    constructor({ username, password }, options) {
        super();
        if (username === undefined || password === undefined || username === null || password === null) {
            throw new Error(`No Basic credentials for Thing`);
        }
        this.username = username;
        this.password = password;
        this.options = options;
    }
    async sign(request) {
        const result = request.clone();
        let headerName = "authorization";
        if (this.options !== undefined && this.options.in === "header" && this.options.name !== undefined) {
            headerName = this.options.name;
        }
        result.headers.set(headerName, "Basic " + Buffer.from(this.username + ":" + this.password).toString("base64"));
        return result;
    }
}
exports.BasicCredential = BasicCredential;
class BearerCredential extends Credential {
    constructor({ token }, options) {
        super();
        if (token === undefined || token === null) {
            throw new Error(`No Bearer credentials for Thing`);
        }
        this.token = token;
        this.options = options;
    }
    async sign(request) {
        const result = request.clone();
        let headerName = "authorization";
        if (this.options.in === "header" && this.options.name !== undefined) {
            headerName = this.options.name;
        }
        result.headers.set(headerName, "Bearer " + this.token);
        return result;
    }
}
exports.BearerCredential = BearerCredential;
class BasicKeyCredential extends Credential {
    constructor({ apiKey }, options) {
        super();
        if (apiKey === undefined || apiKey === null) {
            throw new Error(`No API key credentials for Thing`);
        }
        this.apiKey = apiKey;
        this.options = options;
    }
    async sign(request) {
        const result = request.clone();
        let headerName = "authorization";
        if (this.options.in === "header" && this.options.name !== undefined) {
            headerName = this.options.name;
        }
        result.headers.append(headerName, this.apiKey);
        return result;
    }
}
exports.BasicKeyCredential = BasicKeyCredential;
class OAuthCredential extends Credential {
    constructor(token, refresh) {
        super();
        this.token = token;
        this.refresh = refresh;
        this.token = token;
    }
    async sign(request) {
        if (this.token instanceof Promise) {
            const tokenRequest = this.token;
            this.token = await tokenRequest;
        }
        let tempRequest = { url: request.url, headers: {} };
        tempRequest = this.token.sign(tempRequest);
        const mergeHeaders = new node_fetch_1.Request(request, tempRequest);
        return mergeHeaders;
    }
    async refreshToken() {
        if (this.token instanceof Promise) {
            throw new Error("Uninitialized token. You have to call sing before refresh");
        }
        let newToken;
        if (this.refresh) {
            newToken = await this.refresh();
        }
        else {
            newToken = await this.token.refresh();
        }
        return new OAuthCredential(newToken, this.refresh);
    }
}
exports.OAuthCredential = OAuthCredential;
class TuyaCustomBearer extends Credential {
    constructor(credentials, scheme) {
        super();
        this.key = credentials.key;
        this.secret = credentials.secret;
        this.baseUri = scheme.baseUri;
    }
    async sign(request) {
        const isTokenExpired = this.isTokenExpired();
        if (this.token === undefined || this.token === "" || isTokenExpired)
            await this.requestAndRefreshToken(isTokenExpired);
        const url = request.url;
        const body = request.body?.read().toString();
        const method = request.method;
        const headers = this.getHeaders(true, request.headers.raw(), body, url, method);
        Object.assign(headers, request.headers.raw());
        return new node_fetch_1.Request(url, { method, body: body !== "" ? body : undefined, headers });
    }
    async requestAndRefreshToken(refresh) {
        const headers = this.getHeaders(false, {});
        const request = {
            headers,
            method: "GET",
        };
        let url = `${this.baseUri}/token?grant_type=1`;
        if (refresh) {
            url = `${this.baseUri}/token/${this.refreshToken}`;
        }
        const data = await (await (0, node_fetch_1.default)(url, request)).json();
        const success = data.success ?? false;
        if (success) {
            this.token = data.result?.access_token;
            this.refreshToken = data.result?.refresh_token;
            const expireTime = data.result?.expire_time;
            if (expireTime != null) {
                this.expireTime = new Date(Date.now() + expireTime * 1000);
            }
        }
        else {
            throw new Error("token fetch failed");
        }
    }
    getHeaders(NormalRequest, headers, body, url, method) {
        const requestTime = Date.now().toString();
        const replaceUri = this.baseUri.replace("/v1.0", "");
        const _url = url?.replace(replaceUri, "");
        const sign = this.requestSign(NormalRequest, requestTime, body, _url, method);
        return {
            t: requestTime,
            client_id: this.key,
            sign_method: "HMAC-SHA256",
            sign,
            access_token: this.token ?? "",
        };
    }
    requestSign(NormalRequest, requestTime, body, path = "", method) {
        const bodyHash = crypto
            .createHash("sha256")
            .update(body ?? "")
            .digest("hex");
        let signUrl = "/v1.0/token?grant_type=1";
        const headerString = "";
        let useToken = "";
        const _method = method ?? "GET";
        if (NormalRequest) {
            useToken = this.token ?? "";
            const pathQuery = queryString.parse(path.split("?")[1]);
            let query = {};
            query = Object.assign(query, pathQuery);
            const sortedQuery = {};
            Object.keys(query)
                .sort()
                .forEach((i) => {
                sortedQuery[i] = query[i];
            });
            const qs = queryString.stringify(sortedQuery);
            signUrl = decodeURIComponent(qs ? `${path.split("?")[0]}?${qs}` : path);
        }
        const endStr = [this.key, useToken, requestTime, [_method, bodyHash, headerString, signUrl].join("\n")].join("");
        const sign = crypto.createHmac("sha256", this.secret).update(endStr).digest("hex").toUpperCase();
        return sign;
    }
    isTokenExpired() {
        return this.expireTime ? Date.now() > this.expireTime.getTime() : false;
    }
}
exports.TuyaCustomBearer = TuyaCustomBearer;
