// @node-wot/binding-http/dist/http-server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const http = __importStar(require("http"));
const https = __importStar(require("https"));
const basic_auth_1 = __importDefault(require("basic-auth"));
const core_1 = require("@node-wot/core");
const oauth_token_validation_1 = __importDefault(require("./oauth-token-validation"));
const slugify_1 = __importDefault(require("slugify"));
const find_my_way_1 = __importDefault(require("find-my-way"));
const things_1 = __importDefault(require("./routes/things"));
const thing_description_1 = __importDefault(require("./routes/thing-description"));
const property_1 = __importDefault(require("./routes/property"));
const action_1 = __importDefault(require("./routes/action"));
const event_1 = __importDefault(require("./routes/event"));
const properties_1 = __importDefault(require("./routes/properties"));
const property_observe_1 = __importDefault(require("./routes/property-observe"));
const { debug, info, warn, error } = (0, core_1.createLoggers)("binding-http", "http-server");
class HttpServer {
    constructor(config = {}) {
        this.PROPERTY_DIR = "properties";
        this.ACTION_DIR = "actions";
        this.EVENT_DIR = "events";
        this.OBSERVABLE_DIR = "observable";
        this.supportedSecuritySchemes = ["nosec"];
        this.validOAuthClients = /.*/g;
        this.things = new Map();
        this.servient = null;
        this.oAuthValidator = undefined;
        if (typeof config !== "object") {
            throw new Error(`HttpServer requires config object (got ${typeof config})`);
        }
        this.port = this.obtainEnvironmentPortNumber() ?? config.port ?? 8080;
        this.address = config.address;
        this.baseUri = config.baseUri;
        this.urlRewrite = config.urlRewrite;
        this.middleware = config.middleware;
        this.devFriendlyUri = config.devFriendlyUri ?? true;
        const router = (0, find_my_way_1.default)({
            ignoreTrailingSlash: true,
            defaultRoute(req, res) {
                const pathname = req.url;
                if (config.urlRewrite) {
                    const entryUrl = pathname;
                    const internalUrl = config.urlRewrite[entryUrl ?? "/"];
                    if (internalUrl) {
                        req.url = internalUrl;
                        router.lookup(req, res, this);
                        debug("[binding-http]", `URL "${entryUrl}" has been rewritten to "${pathname}"`);
                        return;
                    }
                }
                res.writeHead(404);
                res.end("Not Found");
            },
        });
        this.router = router;
        this.router.get("/", things_1.default);
        this.router.get("/:thing", thing_description_1.default);
        this.router.on(["GET", "HEAD", "OPTIONS"], "/:thing/" + this.PROPERTY_DIR, properties_1.default);
        this.router.on(["GET", "PUT", "HEAD", "OPTIONS"], "/:thing/" + this.PROPERTY_DIR + "/:property", property_1.default);
        this.router.on(["GET", "HEAD", "OPTIONS"], "/:thing/" + this.PROPERTY_DIR + "/:property/" + this.OBSERVABLE_DIR, property_observe_1.default);
        this.router.on(["POST", "OPTIONS"], "/:thing/" + this.ACTION_DIR + "/:action", action_1.default);
        this.router.on(["GET", "HEAD", "OPTIONS"], "/:thing/" + this.EVENT_DIR + "/:event", event_1.default);
        if (config.serverKey != null && config.serverCert != null) {
            const options = {};
            options.key = fs.readFileSync(config.serverKey);
            options.cert = fs.readFileSync(config.serverCert);
            this.scheme = "https";
            this.server = https.createServer(options, (req, res) => {
                if (this.middleware) {
                    this.middleware(req, res, () => {
                        this.handleRequest(req, res);
                    });
                }
                else {
                    this.handleRequest(req, res);
                }
            });
        }
        else {
            this.scheme = "http";
            this.server = http.createServer((req, res) => {
                if (this.middleware) {
                    this.middleware(req, res, () => {
                        this.handleRequest(req, res);
                    });
                }
                else {
                    this.handleRequest(req, res);
                }
            });
        }
        if (config.security) {
            if (config.security.length > 1) {
                this.supportedSecuritySchemes = [];
            }
            for (const securityScheme of config.security) {
                switch (securityScheme.scheme) {
                    case "nosec":
                    case "basic":
                    case "digest":
                    case "bearer":
                        break;
                    case "oauth2":
                        {
                            const oAuthConfig = securityScheme;
                            this.validOAuthClients = new RegExp(oAuthConfig.allowedClients ?? ".*");
                            this.oAuthValidator = (0, oauth_token_validation_1.default)(oAuthConfig.method);
                        }
                        break;
                    default:
                        throw new Error(`HttpServer does not support security scheme '${securityScheme.scheme}`);
                }
                this.supportedSecuritySchemes.push(securityScheme.scheme);
            }
        }
    }
    obtainEnvironmentPortNumber() {
        for (const portVariable of ["WOT_PORT", "PORT"]) {
            const environmentValue = process.env[portVariable];
            if (environmentValue == null) {
                continue;
            }
            const parsedPort = parseInt(environmentValue);
            if (isNaN(parsedPort)) {
                debug(`Ignoring environment variable ${portVariable} because it is not an integer.`);
                continue;
            }
            info(`HttpServer Port Overridden to ${parsedPort} by Environment Variable ${portVariable}`);
            return parsedPort;
        }
        return undefined;
    }
    start(servient) {
        info(`HttpServer starting on ${this.address !== undefined ? this.address + " " : ""}port ${this.port}`);
        return new Promise((resolve, reject) => {
            this.servient = servient;
            this.server.setTimeout(60 * 60 * 1000, () => {
                debug(`HttpServer on port ${this.getPort()} timed out connection`);
            });
            this.server.keepAliveTimeout = 0;
            this.server.once("error", (err) => {
                reject(err);
            });
            this.server.once("listening", () => {
                this.server.on("error", (err) => {
                    error(`HttpServer on port ${this.port} failed: ${err.message}`);
                });
                resolve();
            });
            this.server.listen(this.port, this.address);
        });
    }
    stop() {
        info(`HttpServer stopping on port ${this.getPort()}`);
        return new Promise((resolve, reject) => {
            this.server.once("error", (err) => {
                reject(err);
            });
            this.server.once("close", () => {
                resolve();
            });
            this.server.close();
        });
    }
    getServer() {
        return this.server;
    }
    getThings() {
        return this.things;
    }
    getPort() {
        const address = this.server?.address();
        if (typeof address === "object") {
            return address?.port ?? -1;
        }
        const port = parseInt(address);
        if (isNaN(port)) {
            return -1;
        }
        return port;
    }
    async expose(thing, tdTemplate = {}) {
        if (this.getPort() !== -1) {
            const paths = [];
            if (this.devFriendlyUri || thing.id == null) {
                let urlPath = (0, slugify_1.default)(thing.title, { lower: true });
                if (this.things.has(urlPath)) {
                    let uniqueUrlPath;
                    let nameClashCnt = 2;
                    do {
                        uniqueUrlPath = urlPath + "_" + nameClashCnt++;
                    } while (this.things.has(uniqueUrlPath));
                    urlPath = uniqueUrlPath;
                }
                this.things.set(urlPath, thing);
                paths.push(urlPath);
                debug("HttpServer on port %d exposes %s as unique '/%s'", this.getPort(), thing.name, urlPath);
            }
            if (thing.id != null) {
                this.things.set(thing.id, thing);
                paths.push(thing.id);
                debug("HttpServer on port %d exposes %s as unique '/%s'", this.getPort(), thing.name, thing.id);
            }
            if (this.scheme === "http" && Object.keys(thing.securityDefinitions).length !== 0) {
                warn(`HTTP Server will attempt to use your security schemes even if you are not using HTTPS.`);
            }
            this.fillSecurityScheme(thing);
            if (this.baseUri !== undefined) {
                for (const path of paths) {
                    info("HttpServer TD hrefs using baseUri %s and path %s", this.baseUri, path);
                    const base = this.baseUri.concat("/", encodeURIComponent(path));
                    this.addEndpoint(thing, tdTemplate, base);
                }
            }
            else {
                for (const address of core_1.Helpers.getAddresses()) {
                    for (const path of paths) {
                        const base = this.scheme + "://" + address + ":" + this.getPort() + "/" + encodeURIComponent(path);
                        info("HttpServer TD hrefs using address %s and path %s", address, path);
                        this.addEndpoint(thing, tdTemplate, base);
                    }
                }
            }
        }
    }
    async destroy(thingId) {
        debug(`HttpServer on port ${this.getPort()} destroying thingId '${thingId}'`);
        for (const [name, thing] of this.things.entries()) {
            if (thing.id === thingId) {
                this.things.delete(name);
                this.things.delete(thingId);
                info(`HttpServer successfully destroyed '${thing.title}'`);
                return true;
            }
        }
        info(`HttpServer failed to destroy thing with thingId '${thingId}'`);
        return false;
    }
    addUrlRewriteEndpoints(form, forms) {
        if (this.urlRewrite != null) {
            for (const [inUri, toUri] of Object.entries(this.urlRewrite)) {
                const endsWithToUri = form.href.endsWith(toUri);
                if (endsWithToUri) {
                    const form2 = core_1.Helpers.structuredClone(form);
                    form2.href = form2.href.substring(0, form.href.lastIndexOf(toUri)) + inUri;
                    forms.push(form2);
                    debug(`HttpServer on port ${this.getPort()} assigns urlRewrite '${form2.href}' for '${form.href}'`);
                }
            }
        }
    }
    addEndpoint(thing, tdTemplate, base) {
        for (const type of core_1.ContentSerdes.get().getOfferedMediaTypes()) {
            const properties = Object.values(thing.properties);
            let allReadOnly = true;
            let allWriteOnly = true;
            for (const property of properties) {
                const readOnly = property.readOnly ?? false;
                if (!readOnly) {
                    allReadOnly = false;
                }
                const writeOnly = property.writeOnly ?? false;
                if (!writeOnly) {
                    allWriteOnly = false;
                }
            }
            if (properties.length > 0) {
                const href = base + "/" + this.PROPERTY_DIR;
                const form = new core_1.Form(href, type);
                if (allReadOnly && !allWriteOnly) {
                    form.op = ["readallproperties", "readmultipleproperties"];
                }
                else if (allWriteOnly && !allReadOnly) {
                    form.op = ["writeallproperties", "writemultipleproperties"];
                }
                else {
                    form.op = [
                        "readallproperties",
                        "readmultipleproperties",
                        "writeallproperties",
                        "writemultipleproperties",
                    ];
                }
                thing.forms ??= [];
                thing.forms.push(form);
                this.addUrlRewriteEndpoints(form, thing.forms);
            }
            for (const [propertyName, property] of Object.entries(thing.properties)) {
                const propertyNamePattern = core_1.Helpers.updateInteractionNameWithUriVariablePattern(propertyName, property.uriVariables, thing.uriVariables);
                const href = base + "/" + this.PROPERTY_DIR + "/" + propertyNamePattern;
                const form = new core_1.Form(href, type);
                core_1.ProtocolHelpers.updatePropertyFormWithTemplate(form, (tdTemplate.properties?.[propertyName] ?? {}));
                const readOnly = property.readOnly ?? false;
                const writeOnly = property.writeOnly ?? false;
                if (readOnly) {
                    form.op = ["readproperty"];
                    const hform = form;
                    hform["htv:methodName"] ??= "GET";
                }
                else if (writeOnly) {
                    form.op = ["writeproperty"];
                    const hform = form;
                    hform["htv:methodName"] ??= "PUT";
                }
                else {
                    form.op = ["readproperty", "writeproperty"];
                }
                property.forms.push(form);
                debug(`HttpServer on port ${this.getPort()} assigns '${href}' to Property '${propertyName}'`);
                this.addUrlRewriteEndpoints(form, property.forms);
                if (property.observable === true) {
                    const href = base +
                        "/" +
                        this.PROPERTY_DIR +
                        "/" +
                        encodeURIComponent(propertyName) +
                        "/" +
                        this.OBSERVABLE_DIR;
                    const form = new core_1.Form(href, type);
                    form.op = ["observeproperty", "unobserveproperty"];
                    form.subprotocol = "longpoll";
                    property.forms.push(form);
                    debug(`HttpServer on port ${this.getPort()} assigns '${href}' to observable Property '${propertyName}'`);
                    this.addUrlRewriteEndpoints(form, property.forms);
                }
            }
            for (const [actionName, action] of Object.entries(thing.actions)) {
                const actionNamePattern = core_1.Helpers.updateInteractionNameWithUriVariablePattern(actionName, action.uriVariables, thing.uriVariables);
                const href = base + "/" + this.ACTION_DIR + "/" + actionNamePattern;
                const form = new core_1.Form(href, type);
                core_1.ProtocolHelpers.updateActionFormWithTemplate(form, (tdTemplate.actions?.[actionName] ?? {}));
                form.op = ["invokeaction"];
                const hform = form;
                hform["htv:methodName"] ??= "POST";
                action.forms.push(form);
                debug(`HttpServer on port ${this.getPort()} assigns '${href}' to Action '${actionName}'`);
                this.addUrlRewriteEndpoints(form, action.forms);
            }
            for (const [eventName, event] of Object.entries(thing.events)) {
                const eventNamePattern = core_1.Helpers.updateInteractionNameWithUriVariablePattern(eventName, event.uriVariables, thing.uriVariables);
                const href = base + "/" + this.EVENT_DIR + "/" + eventNamePattern;
                const form = new core_1.Form(href, type);
                core_1.ProtocolHelpers.updateEventFormWithTemplate(form, (tdTemplate.events?.[eventName] ?? {}));
                form.subprotocol = "longpoll";
                form.op = ["subscribeevent", "unsubscribeevent"];
                event.forms.push(form);
                debug(`HttpServer on port ${this.getPort()} assigns '${href}' to Event '${eventName}'`);
                this.addUrlRewriteEndpoints(form, event.forms);
            }
        }
    }
    async checkCredentials(thing, req) {
        debug(`HttpServer on port ${this.getPort()} checking credentials for '${thing.id}'`);
        if (this.servient === null) {
            throw new Error("Servient not set");
        }
        const credentials = this.servient.retrieveCredentials(thing.id);
        const selected = core_1.Helpers.toStringArray(thing.security)[0];
        const thingSecurityScheme = thing.securityDefinitions[selected];
        debug(`Verifying credentials with security scheme '${thingSecurityScheme.scheme}'`);
        switch (thingSecurityScheme.scheme) {
            case "nosec":
                return true;
            case "basic": {
                const basic = (0, basic_auth_1.default)(req);
                if (basic === undefined)
                    return false;
                if (credentials == null || credentials.length === 0)
                    return false;
                const basicCredentials = credentials;
                return basicCredentials.some((cred) => basic.name === cred.username && basic.pass === cred.password);
            }
            case "digest":
                return false;
            case "oauth2": {
                const oAuthScheme = thing.securityDefinitions[thing.security[0]];
                const scopes = core_1.Helpers.toStringArray(oAuthScheme.scopes);
                let valid = false;
                if (!this.oAuthValidator) {
                    throw new Error("OAuth validator not set. Cannot validate request.");
                }
                try {
                    valid = await this.oAuthValidator.validate(req, scopes, this.validOAuthClients);
                }
                catch (err) {
                    error("OAuth authorization error; sending unauthorized response error");
                    error("this was possibly caused by a misconfiguration of the server");
                    error(`${err}`);
                }
                return valid;
            }
            case "Bearer": {
                if (req.headers.authorization === undefined)
                    return false;
                const auth = req.headers.authorization.split(" ");
                if (auth.length !== 2 || auth[0] !== "Bearer")
                    return false;
                if (credentials == null || credentials.length === 0)
                    return false;
                const bearerCredentials = credentials;
                return bearerCredentials.some((cred) => cred.token === auth[1]);
            }
            default:
                return false;
        }
    }
    fillSecurityScheme(thing) {
        if (thing.security.length > 0) {
            const securityScheme = core_1.Helpers.toStringArray(thing.security)[0];
            const secCandidate = Object.keys(thing.securityDefinitions).find((key) => {
                return key === securityScheme;
            });
            if (secCandidate == null) {
                throw new Error("Security scheme not found in thing security definitions. Thing security definitions: " +
                    Object.keys(thing.securityDefinitions).join(", "));
            }
            const isSupported = this.supportedSecuritySchemes.find((supportedScheme) => {
                const thingScheme = thing.securityDefinitions[secCandidate].scheme;
                return thingScheme === supportedScheme.toLocaleLowerCase();
            });
            if (isSupported == null) {
                throw new Error("Servient does not support thing security schemes. Current scheme supported: " +
                    this.supportedSecuritySchemes.join(", "));
            }
            return;
        }
        if (Object.keys(thing.securityDefinitions ?? {}).length === 0) {
            thing.securityDefinitions = {
                [this.supportedSecuritySchemes[0]]: { scheme: this.supportedSecuritySchemes[0] },
            };
            thing.security = [this.supportedSecuritySchemes[0]];
            return;
        }
        const secCandidate = Object.keys(thing.securityDefinitions).find((key) => {
            let scheme = thing.securityDefinitions[key].scheme;
            scheme = scheme === "oauth2" ? scheme.split("2")[0] : scheme;
            return this.supportedSecuritySchemes.includes(scheme.toLocaleLowerCase());
        });
        if (secCandidate == null) {
            throw new Error("Servient does not support any of thing security schemes. Current scheme supported: " +
                this.supportedSecuritySchemes.join(",") +
                " thing security schemes: " +
                Object.values(thing.securityDefinitions)
                    .map((schemeDef) => schemeDef.scheme)
                    .join(", "));
        }
        const selectedSecurityScheme = thing.securityDefinitions[secCandidate];
        thing.securityDefinitions = {};
        thing.securityDefinitions[secCandidate] = selectedSecurityScheme;
        thing.security = [secCandidate];
    }
    async handleRequest(req, res) {
        const requestUri = new URL(req.url ?? "", `${this.scheme}://${req.headers.host}`);
        debug(`HttpServer on port ${this.getPort()} received '${req.method} ${requestUri.pathname}' from ${core_1.Helpers.toUriLiteral(req.socket.remoteAddress)}:${req.socket.remotePort}`);
        res.on("finish", () => {
            debug(`HttpServer on port ${this.getPort()} replied with '${res.statusCode}' to ${core_1.Helpers.toUriLiteral(req.socket.remoteAddress)}:${req.socket.remotePort}`);
        });
        const contentTypeHeader = req.headers["content-type"];
        let contentType = Array.isArray(contentTypeHeader) ? contentTypeHeader[0] : contentTypeHeader;
        if (req.method === "PUT" || req.method === "POST") {
            if (!contentType) {
                warn(`HttpServer on port ${this.getPort()} received no Content-Type from ${core_1.Helpers.toUriLiteral(req.socket.remoteAddress)}:${req.socket.remotePort}`);
                contentType = core_1.ContentSerdes.DEFAULT;
            }
            else if (core_1.ContentSerdes.get().getSupportedMediaTypes().indexOf(core_1.ContentSerdes.getMediaType(contentType)) < 0) {
                res.writeHead(415);
                res.end("Unsupported Media Type");
                return;
            }
        }
        this.router.lookup(req, res, this);
    }
}
exports.default = HttpServer;
