// @node-wot/core/dist/consumed-thing.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _ConsumedThing_servient, _ConsumedThing_clients;
Object.defineProperty(exports, "__esModule", { value: true });
const thing_description_1 = require("./thing-description");
const helpers_1 = __importDefault(require("./helpers"));
const content_type_1 = __importDefault(require("content-type"));
const content_serdes_1 = __importDefault(require("./content-serdes"));
const UriTemplate = require("uritemplate");
const interaction_output_1 = require("./interaction-output");
const logger_1 = require("./logger");
const { debug, warn } = (0, logger_1.createLoggers)("core", "consumed-thing");
var Affordance;
(function (Affordance) {
    Affordance[Affordance["PropertyAffordance"] = 0] = "PropertyAffordance";
    Affordance[Affordance["ActionAffordance"] = 1] = "ActionAffordance";
    Affordance[Affordance["EventAffordance"] = 2] = "EventAffordance";
})(Affordance || (Affordance = {}));
class ConsumedThingProperty extends thing_description_1.ThingProperty {
    constructor() {
        super();
    }
}
class ConsumedThingAction extends thing_description_1.ThingAction {
    constructor() {
        super();
    }
}
class ConsumedThingEvent extends thing_description_1.ThingEvent {
    constructor() {
        super();
    }
}
class InternalSubscription {
    constructor(thing, name, client) {
        this.thing = thing;
        this.name = name;
        this.client = client;
        this.active = true;
    }
}
function handleUriVariables(thing, ti, form, options) {
    const ut = UriTemplate.parse(form.href);
    const uriVariables = helpers_1.default.parseInteractionOptions(thing, ti, options).uriVariables;
    const updatedHref = ut.expand(uriVariables ?? {});
    if (updatedHref !== form.href) {
        const updForm = { ...form };
        updForm.href = updatedHref;
        form = updForm;
        debug(`ConsumedThing '${thing.title}' update form URI to ${form.href}`);
    }
    return form;
}
class InternalPropertySubscription extends InternalSubscription {
    constructor(thing, name, client, form) {
        super(thing, name, client);
        this.form = form;
        this.active = false;
        const index = this.thing.properties?.[name].forms.indexOf(form);
        if (index === undefined || index < 0) {
            throw new Error(`Could not find form ${form.href} in property ${name}`);
        }
        this.formIndex = index;
    }
    async stop(options) {
        await this.unobserveProperty(options);
        this.thing["observedProperties"].delete(this.name);
    }
    async unobserveProperty(options = {}) {
        const tp = this.thing.properties[this.name];
        if (tp == null) {
            throw new Error(`ConsumedThing '${this.thing.title}' does not have property ${this.name}`);
        }
        options.formIndex ??= this.matchingUnsubscribeForm();
        const { form } = this.thing.getClientFor(tp.forms, "unobserveproperty", Affordance.PropertyAffordance, options);
        if (form == null) {
            throw new Error(`ConsumedThing '${this.thing.title}' did not get suitable form`);
        }
        const formWithoutURIvariables = handleUriVariables(this.thing, tp, form, options);
        debug(`ConsumedThing '${this.thing.title}' unobserving to ${form.href}`);
        await this.client.unlinkResource(formWithoutURIvariables);
        this.active = false;
    }
    matchingUnsubscribeForm() {
        const refForm = this.thing.properties[this.name].forms[this.formIndex];
        if (Array.isArray(refForm.op) && refForm.op.includes("unobserveproperty")) {
            return this.formIndex;
        }
        const bestFormMatch = this.findFormIndexWithScoring(this.formIndex, this.thing.properties[this.name].forms, "unobserveproperty");
        if (bestFormMatch === -1) {
            throw new Error(`Could not find matching form for unsubscribe`);
        }
        return bestFormMatch;
    }
    findFormIndexWithScoring(formIndex, forms, operation) {
        const refForm = forms[formIndex];
        let maxScore = 0;
        let maxScoreIndex = -1;
        for (let i = 0; i < forms.length; i++) {
            let score = 0;
            const form = forms[i];
            if (form.op === operation || (form?.op?.includes(operation) === true && Array.isArray(form.op) === true)) {
                score += 1;
            }
            if (new URL(form.href).origin === new URL(refForm.href).origin) {
                score += 1;
            }
            if (form.contentType === refForm.contentType) {
                score += 1;
            }
            if (score > maxScore) {
                maxScore = score;
                maxScoreIndex = i;
            }
        }
        return maxScoreIndex;
    }
}
function findFormIndexWithScoring(formIndex, forms, operation) {
    const refForm = forms[formIndex];
    let maxScore = 0;
    let maxScoreIndex = -1;
    for (let i = 0; i < forms.length; i++) {
        let score = 0;
        const form = forms[i];
        if (form.op === operation || (form?.op?.includes(operation) === true && Array.isArray(form.op) === true)) {
            score += 1;
        }
        if (new URL(form.href).origin === new URL(refForm.href).origin) {
            score += 1;
        }
        if (form.contentType === refForm.contentType) {
            score += 1;
        }
        if (score > maxScore) {
            maxScore = score;
            maxScoreIndex = i;
        }
    }
    return maxScoreIndex;
}
class InternalEventSubscription extends InternalSubscription {
    constructor(thing, name, client, form) {
        super(thing, name, client);
        this.form = form;
        const index = this.thing.events?.[name].forms.indexOf(form);
        if (index === undefined || index < 0) {
            throw new Error(`Could not find form ${form.href} in event ${name}`);
        }
        this.formIndex = index;
    }
    async stop(options) {
        await this.unsubscribeEvent(options);
        this.thing["subscribedEvents"].delete(this.name);
    }
    async unsubscribeEvent(options = {}) {
        const te = this.thing.events[this.name];
        if (te == null) {
            throw new Error(`ConsumedThing '${this.thing.title}' does not have event ${this.name}`);
        }
        options.formIndex ??= this.matchingUnsubscribeForm();
        const { form } = this.thing.getClientFor(te.forms, "unsubscribeevent", Affordance.EventAffordance, options);
        if (form == null) {
            throw new Error(`ConsumedThing '${this.thing.title}' did not get suitable form`);
        }
        const formWithoutURIvariables = handleUriVariables(this.thing, te, form, options);
        debug(`ConsumedThing '${this.thing.title}' unsubscribing to ${form.href}`);
        this.client.unlinkResource(formWithoutURIvariables);
        this.active = false;
    }
    matchingUnsubscribeForm() {
        const refForm = this.thing.events[this.name].forms[this.formIndex];
        if (refForm.op == null || (Array.isArray(refForm.op) && refForm.op.includes("unsubscribeevent"))) {
            return this.formIndex;
        }
        const bestFormMatch = findFormIndexWithScoring(this.formIndex, this.thing.events[this.name].forms, "unsubscribeevent");
        if (bestFormMatch === -1) {
            throw new Error(`Could not find matching form for unsubscribe`);
        }
        return bestFormMatch;
    }
}
class ConsumedThing extends thing_description_1.Thing {
    constructor(servient, thingModel) {
        super();
        _ConsumedThing_servient.set(this, void 0);
        _ConsumedThing_clients.set(this, void 0);
        this.subscribedEvents = new Map();
        this.observedProperties = new Map();
        __classPrivateFieldSet(this, _ConsumedThing_servient, servient, "f");
        __classPrivateFieldSet(this, _ConsumedThing_clients, new Map(), "f");
        this.properties = {};
        this.actions = {};
        this.events = {};
        const deepClonedModel = helpers_1.default.structuredClone(thingModel ?? {});
        Object.assign(this, deepClonedModel);
        this.extendInteractions();
    }
    getThingDescription() {
        return JSON.parse(JSON.stringify(this));
    }
    emitEvent(name, data) {
        warn("not implemented");
    }
    extendInteractions() {
        for (const [propertyName, property] of Object.entries(this.properties)) {
            const newProp = helpers_1.default.extend(property, new ConsumedThingProperty());
            this.properties[propertyName] = newProp;
        }
        for (const [actionName, action] of Object.entries(this.actions)) {
            const newAction = helpers_1.default.extend(action, new ConsumedThingAction());
            this.actions[actionName] = newAction;
        }
        for (const [eventName, event] of Object.entries(this.events)) {
            const newEvent = helpers_1.default.extend(event, new ConsumedThingEvent());
            this.events[eventName] = newEvent;
        }
    }
    findForm(forms, op, affordance, schemes, idx) {
        let form;
        for (const f of forms) {
            let fop = "";
            if (f.op !== undefined) {
                fop = f.op;
            }
            else {
                switch (affordance) {
                    case Affordance.PropertyAffordance:
                        fop = ["readproperty", "writeproperty"];
                        break;
                    case Affordance.ActionAffordance:
                        fop = "invokeaction";
                        break;
                    case Affordance.EventAffordance:
                        fop = "subscribeevent";
                        break;
                }
            }
            if (fop.indexOf(op) !== -1 && f.href.indexOf(schemes[idx] + ":") !== -1) {
                form = f;
                break;
            }
        }
        return form;
    }
    getSecuritySchemes(security) {
        const alreadyProcessed = new Map();
        const visitSchemes = (security) => {
            const resolveComboScheme = (combo, name) => {
                if (combo.allOf instanceof Array && combo.oneOf === undefined) {
                    const allOf = visitSchemes(combo.allOf);
                    return {
                        scheme: "combo",
                        allOf,
                    };
                }
                else if (combo.oneOf instanceof Array && combo.allOf === undefined) {
                    const oneOf = visitSchemes(combo.oneOf);
                    return {
                        scheme: "combo",
                        oneOf,
                    };
                }
                else {
                    throw new Error(`Combo SecurityScheme '${name}' is invalid`);
                }
            };
            const scs = [];
            for (const s of security) {
                if (alreadyProcessed.has(s)) {
                    scs.push(alreadyProcessed.get(s));
                    continue;
                }
                alreadyProcessed.set(s, null);
                let ws = this.securityDefinitions[s];
                if (ws?.scheme === "combo") {
                    ws = resolveComboScheme(ws, s);
                }
                if (ws != null) {
                    scs.push(ws);
                    alreadyProcessed.set(s, ws);
                }
            }
            return scs;
        };
        return visitSchemes(security);
    }
    ensureClientSecurity(client, form) {
        if (this.securityDefinitions != null) {
            const logStatement = () => debug(`ConsumedThing '${this.title}' setting credentials for ${client} based on thing security`);
            if (form != null && Array.isArray(form.security) && form.security.length > 0) {
                logStatement();
                client.setSecurity(this.getSecuritySchemes(form.security), __classPrivateFieldGet(this, _ConsumedThing_servient, "f").retrieveCredentials(this.id));
            }
            else if (Array.isArray(this.security) && this.security.length > 0) {
                logStatement();
                client.setSecurity(this.getSecuritySchemes(this.security), __classPrivateFieldGet(this, _ConsumedThing_servient, "f").getCredentials(this.id));
            }
        }
    }
    getClientFor(forms, op, affordance, options) {
        if (forms.length === 0) {
            throw new Error(`ConsumedThing '${this.title}' has no links for this interaction`);
        }
        let form;
        let client;
        if (options?.formIndex !== undefined) {
            debug(`ConsumedThing '${this.title}' asked to use formIndex '${options.formIndex}'`);
            if (options.formIndex >= 0 && options.formIndex < forms.length) {
                form = forms[options.formIndex];
                const scheme = helpers_1.default.extractScheme(form.href);
                if (__classPrivateFieldGet(this, _ConsumedThing_servient, "f").hasClientFor(scheme)) {
                    debug(`ConsumedThing '${this.title}' got client for '${scheme}'`);
                    client = __classPrivateFieldGet(this, _ConsumedThing_servient, "f").getClientFor(scheme);
                    if (!__classPrivateFieldGet(this, _ConsumedThing_clients, "f").get(scheme)) {
                        this.ensureClientSecurity(client, form);
                        __classPrivateFieldGet(this, _ConsumedThing_clients, "f").set(scheme, client);
                    }
                }
                else {
                    throw new Error(`ConsumedThing '${this.title}' missing ClientFactory for '${scheme}'`);
                }
            }
            else {
                throw new Error(`ConsumedThing '${this.title}' missing formIndex '${options.formIndex}'`);
            }
        }
        else {
            const schemes = forms.map((link) => helpers_1.default.extractScheme(link.href));
            const cacheIdx = schemes.findIndex((scheme) => __classPrivateFieldGet(this, _ConsumedThing_clients, "f").has(scheme));
            if (cacheIdx !== -1) {
                debug(`ConsumedThing '${this.title}' chose cached client for '${schemes[cacheIdx]}'`);
                client = __classPrivateFieldGet(this, _ConsumedThing_clients, "f").get(schemes[cacheIdx]);
                form = this.findForm(forms, op, affordance, schemes, cacheIdx);
            }
            else {
                debug(`ConsumedThing '${this.title}' has no client in cache (${cacheIdx})`);
                const srvIdx = schemes.findIndex((scheme) => __classPrivateFieldGet(this, _ConsumedThing_servient, "f").hasClientFor(scheme));
                if (srvIdx === -1)
                    throw new Error(`ConsumedThing '${this.title}' missing ClientFactory for '${schemes}'`);
                client = __classPrivateFieldGet(this, _ConsumedThing_servient, "f").getClientFor(schemes[srvIdx]);
                debug(`ConsumedThing '${this.title}' got new client for '${schemes[srvIdx]}'`);
                __classPrivateFieldGet(this, _ConsumedThing_clients, "f").set(schemes[srvIdx], client);
                form = this.findForm(forms, op, affordance, schemes, srvIdx);
                this.ensureClientSecurity(client, form);
            }
        }
        return { client, form };
    }
    async readProperty(propertyName, options) {
        const tp = this.properties[propertyName];
        if (tp == null) {
            throw new Error(`ConsumedThing '${this.title}' does not have property ${propertyName}`);
        }
        const { client, form } = this.getClientFor(tp.forms, "readproperty", Affordance.PropertyAffordance, options);
        if (form == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable form`);
        }
        if (client == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable client for ${form.href}`);
        }
        debug(`ConsumedThing '${this.title}' reading ${form.href}`);
        const formWithUriVariables = this.handleUriVariables(tp, form, options);
        const content = await client.readResource(formWithUriVariables);
        try {
            return this.handleInteractionOutput(content, formWithUriVariables, tp);
        }
        catch (e) {
            const error = e instanceof Error ? e : new Error(JSON.stringify(e));
            throw new Error(`Error while processing property for ${tp.title}. ${error.message}`);
        }
    }
    handleInteractionOutput(content, form, outputDataSchema) {
        content.type ??= form.contentType ?? "application/json";
        this.checkMediaTypeOrThrow(content, form);
        return new interaction_output_1.InteractionOutput(content, form, outputDataSchema);
    }
    checkMediaTypeOrThrow(content, form) {
        if (form.response != null) {
            const parsedMediaTypeContent = content_type_1.default.parse(content.type);
            const parsedMediaTypeForm = content_type_1.default.parse(form.response.contentType);
            if (parsedMediaTypeContent.type !== parsedMediaTypeForm.type) {
                throw new Error(`Unexpected type '${content.type}' in response. Should be '${form.response.contentType}'`);
            }
        }
    }
    handleActionInteractionOutput(content, form, outputDataSchema, synchronous) {
        content.type ??= form.contentType ?? "application/json";
        this.checkMediaTypeOrThrow(content, form);
        return new interaction_output_1.ActionInteractionOutput(content, form, outputDataSchema, synchronous);
    }
    async _readProperties(propertyNames) {
        const promises = [];
        for (const propertyName of propertyNames) {
            promises.push(this.readProperty(propertyName));
        }
        const output = new Map();
        try {
            const result = await Promise.all(promises);
            let index = 0;
            for (const propertyName of propertyNames) {
                output.set(propertyName, result[index]);
                index++;
            }
            return output;
        }
        catch (err) {
            throw new Error(`ConsumedThing '${this.title}', failed to read properties: ${propertyNames}.\n Error: ${err}`);
        }
    }
    readAllProperties(options) {
        const propertyNames = [];
        for (const [propertyName, property] of Object.entries(this.properties)) {
            const { form } = this.getClientFor(property.forms, "readproperty", Affordance.PropertyAffordance, options);
            if (form != null) {
                propertyNames.push(propertyName);
            }
        }
        return this._readProperties(propertyNames);
    }
    readMultipleProperties(propertyNames, options) {
        return this._readProperties(propertyNames);
    }
    async writeProperty(propertyName, value, options) {
        const tp = this.properties[propertyName];
        if (tp == null) {
            throw new Error(`ConsumedThing '${this.title}' does not have property ${propertyName}`);
        }
        const { client, form } = this.getClientFor(tp.forms, "writeproperty", Affordance.PropertyAffordance, options);
        if (form == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable form`);
        }
        if (client == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable client for ${form.href}`);
        }
        debug(`ConsumedThing '${this.title}' writing ${form.href} with '${value}'`);
        const content = content_serdes_1.default.valueToContent(value, tp, form.contentType);
        const formWithUriVariables = this.handleUriVariables(tp, form, options);
        await client.writeResource(formWithUriVariables, content);
    }
    async writeMultipleProperties(valueMap, options) {
        const promises = [];
        for (const [propertyName, value] of valueMap.entries()) {
            promises.push(this.writeProperty(propertyName, value));
        }
        try {
            await Promise.all(promises);
        }
        catch (err) {
            throw new Error(`ConsumedThing '${this.title}', failed to write multiple properties: ${valueMap}\n Error: ${err}`);
        }
    }
    async invokeAction(actionName, parameter, options) {
        const ta = this.actions[actionName];
        if (ta == null) {
            throw new Error(`ConsumedThing '${this.title}' does not have action ${actionName}`);
        }
        const { client, form } = this.getClientFor(ta.forms, "invokeaction", Affordance.ActionAffordance, options);
        if (form == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable form`);
        }
        if (client == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable client for ${form.href}`);
        }
        debug(`ConsumedThing '${this.title}' invoking ${form.href}${parameter !== undefined ? " with '" + parameter + "'" : ""}`);
        let input;
        if (parameter !== undefined) {
            input = content_serdes_1.default.valueToContent(parameter, ta.input, form.contentType);
        }
        const formWithUriVariables = this.handleUriVariables(ta, form, options);
        const content = await client.invokeResource(formWithUriVariables, input);
        try {
            return this.handleActionInteractionOutput(content, formWithUriVariables, ta.output, ta.synchronous);
        }
        catch (e) {
            const error = e instanceof Error ? e : new Error(JSON.stringify(e));
            throw new Error(`Error while processing action for ${ta.title}. ${error.message}`);
        }
    }
    async observeProperty(name, listener, errorListener, options) {
        const tp = this.properties[name];
        if (tp == null) {
            throw new Error(`ConsumedThing '${this.title}' does not have property ${name}`);
        }
        const { client, form } = this.getClientFor(tp.forms, "observeproperty", Affordance.PropertyAffordance, options);
        if (form == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable form`);
        }
        if (client == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable client for ${form.href}`);
        }
        if (this.observedProperties.has(name)) {
            throw new Error(`ConsumedThing '${this.title}' has already a function subscribed to ${name}. You can only observe once`);
        }
        debug(`ConsumedThing '${this.title}' observing to ${form.href}`);
        const formWithoutURITemplates = this.handleUriVariables(tp, form, options);
        await client.subscribeResource(formWithoutURITemplates, (content) => {
            try {
                listener(this.handleInteractionOutput(content, form, tp));
            }
            catch (e) {
                const error = e instanceof Error ? e : new Error(JSON.stringify(e));
                warn(`Error while processing observe property for ${tp.title}. ${error.message}`);
                warn(e);
            }
        }, (err) => {
            errorListener?.(err);
        }, () => {
        });
        const subscription = new InternalPropertySubscription(this, name, client, form);
        this.observedProperties.set(name, subscription);
        return subscription;
    }
    async subscribeEvent(name, listener, errorListener, options) {
        const te = this.events[name];
        if (te == null) {
            throw new Error(`ConsumedThing '${this.title}' does not have event ${name}`);
        }
        const { client, form } = this.getClientFor(te.forms, "subscribeevent", Affordance.EventAffordance, options);
        if (form == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable form`);
        }
        if (client == null) {
            throw new Error(`ConsumedThing '${this.title}' did not get suitable client for ${form.href}`);
        }
        if (this.subscribedEvents.has(name)) {
            throw new Error(`ConsumedThing '${this.title}' has already a function subscribed to ${name}. You can only subscribe once`);
        }
        debug(`ConsumedThing '${this.title}' subscribing to ${form.href}`);
        const formWithoutURITemplates = this.handleUriVariables(te, form, options);
        await client.subscribeResource(formWithoutURITemplates, (content) => {
            try {
                listener(this.handleInteractionOutput(content, form, te.data));
            }
            catch (e) {
                const error = e instanceof Error ? e : new Error(JSON.stringify(e));
                warn(`Error while processing event for ${te.title}. ${error.message}`);
                warn(e);
            }
        }, (err) => {
            errorListener?.(err);
        }, () => {
        });
        const subscription = new InternalEventSubscription(this, name, client, form);
        this.subscribedEvents.set(name, subscription);
        return subscription;
    }
    handleUriVariables(ti, form, options) {
        return handleUriVariables(this, ti, form, options);
    }
}
_ConsumedThing_servient = new WeakMap(), _ConsumedThing_clients = new WeakMap();
exports.default = ConsumedThing;
