// @node-wot/binding-mqtt/dist/mqtt-broker-server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const mqtt = __importStar(require("mqtt"));
const url = __importStar(require("url"));
const aedes_1 = require("aedes");
const net = __importStar(require("net"));
const tls = __importStar(require("tls"));
const mqtt_1 = require("./mqtt");
const core_1 = require("@node-wot/core");
const stream_1 = require("stream");
const util_1 = require("./util");
const { info, debug, error, warn } = (0, core_1.createLoggers)("binding-mqtt", "mqtt-broker-server");
class MqttBrokerServer {
    static brokerIsInitialized(broker) {
        if (broker === undefined) {
            throw new Error(`Broker not initialized. You need to start the ${MqttBrokerServer.name} before you can expose things.`);
        }
    }
    constructor(config) {
        this.scheme = "mqtt";
        this.ACTION_SEGMENT_LENGTH = 3;
        this.PROPERTY_SEGMENT_LENGTH = 4;
        this.THING_NAME_SEGMENT_INDEX = 0;
        this.INTERACTION_TYPE_SEGMENT_INDEX = 1;
        this.INTERACTION_NAME_SEGMENT_INDEX = 2;
        this.INTERACTION_EXT_SEGMENT_INDEX = 3;
        this.defaults = { uri: "mqtt://localhost:1883" };
        this.port = -1;
        this.address = undefined;
        this.things = new Map();
        this.config = config ?? this.defaults;
        this.config.uri = this.config.uri ?? this.defaults.uri;
        if (config.uri.indexOf("://") === -1) {
            config.uri = this.scheme + "://" + config.uri;
        }
        this.brokerURI = config.uri;
    }
    async expose(thing) {
        if (this.broker === undefined) {
            return;
        }
        let name = thing.title;
        if (this.things.has(name)) {
            const suffix = name.match(/.+_([0-9]+)$/);
            if (suffix !== null) {
                name = name.slice(0, -suffix[1].length) + (1 + parseInt(suffix[1]));
            }
            else {
                name = name + "_2";
            }
        }
        debug(`MqttBrokerServer at ${this.brokerURI} exposes '${thing.title}' as unique '${name}/*'`);
        this.things.set(name, thing);
        for (const propertyName of Object.keys(thing.properties)) {
            this.exposeProperty(name, propertyName, thing);
        }
        for (const actionName of Object.keys(thing.actions)) {
            this.exposeAction(name, actionName, thing);
        }
        for (const eventName of Object.keys(thing.events)) {
            this.exposeEvent(name, eventName, thing);
        }
        this.broker.on("message", this.handleMessage.bind(this));
        this.broker.publish(name, JSON.stringify(thing.getThingDescription()), { retain: true });
    }
    exposeProperty(name, propertyName, thing) {
        MqttBrokerServer.brokerIsInitialized(this.broker);
        const topic = encodeURIComponent(name) + "/properties/" + encodeURIComponent(propertyName);
        const property = thing.properties[propertyName];
        const writeOnly = property.writeOnly ?? false;
        if (!writeOnly) {
            const href = this.brokerURI + "/" + topic;
            const form = new core_1.Form(href, core_1.ContentSerdes.DEFAULT);
            form.op = ["readproperty", "observeproperty", "unobserveproperty"];
            property.forms.push(form);
            debug(`MqttBrokerServer at ${this.brokerURI} assigns '${href}' to property '${propertyName}'`);
            const observeListener = async (content) => {
                debug(`MqttBrokerServer at ${this.brokerURI} publishing to Property topic '${propertyName}' `);
                const buffer = await content.toBuffer();
                if (this.broker === undefined) {
                    warn(`MqttBrokerServer at ${this.brokerURI} has no client to publish to. Probably it was closed.`);
                    return;
                }
                this.broker.publish(topic, buffer);
            };
            thing.handleObserveProperty(propertyName, observeListener, { formIndex: property.forms.length - 1 });
        }
        const readOnly = property.readOnly ?? false;
        if (!readOnly) {
            const href = this.brokerURI + "/" + topic + "/writeproperty";
            this.broker.subscribe(topic + "/writeproperty");
            const form = new core_1.Form(href, core_1.ContentSerdes.DEFAULT);
            form.op = ["writeproperty"];
            thing.properties[propertyName].forms.push(form);
            debug(`MqttBrokerServer at ${this.brokerURI} assigns '${href}' to property '${propertyName}'`);
        }
    }
    exposeAction(name, actionName, thing) {
        MqttBrokerServer.brokerIsInitialized(this.broker);
        const topic = encodeURIComponent(name) + "/actions/" + encodeURIComponent(actionName);
        this.broker.subscribe(topic);
        const href = this.brokerURI + "/" + topic;
        const form = new core_1.Form(href, core_1.ContentSerdes.DEFAULT);
        form.op = ["invokeaction"];
        thing.actions[actionName].forms.push(form);
        debug(`MqttBrokerServer at ${this.brokerURI} assigns '${href}' to Action '${actionName}'`);
    }
    exposeEvent(name, eventName, thing) {
        const topic = encodeURIComponent(name) + "/events/" + encodeURIComponent(eventName);
        const event = thing.events[eventName];
        const href = this.brokerURI + "/" + topic;
        const form = new mqtt_1.MqttForm(href, core_1.ContentSerdes.DEFAULT);
        form["mqv:qos"] = "2";
        form.op = ["subscribeevent", "unsubscribeevent"];
        event.forms.push(form);
        debug(`MqttBrokerServer at ${this.brokerURI} assigns '${href}' to Event '${eventName}'`);
        const eventListener = async (content) => {
            if (this.broker === undefined) {
                warn(`MqttBrokerServer at ${this.brokerURI} has no client to publish to. Probably it was closed.`);
                return;
            }
            if (content == null) {
                warn(`MqttBrokerServer on port ${this.getPort()} cannot process data for Event ${eventName}`);
                thing.handleUnsubscribeEvent(eventName, eventListener, { formIndex: event.forms.length - 1 });
                return;
            }
            debug(`MqttBrokerServer at ${this.brokerURI} publishing to Event topic '${eventName}' `);
            const buffer = await content.toBuffer();
            this.broker.publish(topic, buffer, { retain: form["mqv:retain"], qos: (0, util_1.mapQoS)(form["mqv:qos"]) });
        };
        thing.handleSubscribeEvent(eventName, eventListener, { formIndex: event.forms.length - 1 });
    }
    handleMessage(receivedTopic, rawPayload, packet) {
        const segments = receivedTopic.split("/");
        let payload;
        if (rawPayload instanceof Buffer) {
            payload = rawPayload;
        }
        else if (typeof rawPayload === "string") {
            payload = Buffer.from(rawPayload);
        }
        else {
            warn(`MqttBrokerServer on port ${this.getPort()} received unexpected payload type`);
            return;
        }
        if (segments.length === this.ACTION_SEGMENT_LENGTH) {
            debug(`MqttBrokerServer at ${this.brokerURI} received message for '${receivedTopic}'`);
            const thing = this.things.get(segments[this.THING_NAME_SEGMENT_INDEX]);
            if (thing != null) {
                if (segments[this.INTERACTION_TYPE_SEGMENT_INDEX] === "actions") {
                    const action = thing.actions[segments[this.INTERACTION_NAME_SEGMENT_INDEX]];
                    if (action != null) {
                        this.handleAction(action, packet, payload, segments, thing);
                        return;
                    }
                }
            }
        }
        else if (segments.length === this.PROPERTY_SEGMENT_LENGTH &&
            segments[this.INTERACTION_EXT_SEGMENT_INDEX] === "writeproperty") {
            const thing = this.things.get(segments[this.THING_NAME_SEGMENT_INDEX]);
            if (thing != null) {
                if (segments[this.INTERACTION_TYPE_SEGMENT_INDEX] === "properties") {
                    const property = thing.properties[segments[this.INTERACTION_NAME_SEGMENT_INDEX]];
                    if (property != null) {
                        this.handlePropertyWrite(property, packet, payload, segments, thing);
                    }
                }
            }
            return;
        }
        warn(`MqttBrokerServer at ${this.brokerURI} received message for invalid topic '${receivedTopic}'`);
    }
    handleAction(action, packet, payload, segments, thing) {
        const contentType = packet?.properties?.contentType ?? core_1.ContentSerdes.DEFAULT;
        const options = {
            formIndex: core_1.ProtocolHelpers.findRequestMatchingFormIndex(action.forms, this.scheme, this.brokerURI, contentType),
        };
        const formContentType = action.forms[options.formIndex].contentType ?? core_1.ContentSerdes.DEFAULT;
        const inputContent = new core_1.Content(formContentType, stream_1.Readable.from(payload));
        thing
            .handleInvokeAction(segments[this.INTERACTION_NAME_SEGMENT_INDEX], inputContent, options)
            .then((output) => {
            if (output != null) {
                warn(`MqttBrokerServer at ${this.brokerURI} cannot return output '${segments[this.INTERACTION_NAME_SEGMENT_INDEX]}'`);
            }
        })
            .catch((err) => {
            error(`MqttBrokerServer at ${this.brokerURI} got error on invoking '${segments[this.INTERACTION_NAME_SEGMENT_INDEX]}': ${err.message}`);
        });
    }
    handlePropertyWrite(property, packet, payload, segments, thing) {
        const readOnly = property.readOnly ?? false;
        if (!readOnly) {
            const contentType = packet?.properties?.contentType ?? core_1.ContentSerdes.DEFAULT;
            const options = {
                formIndex: core_1.ProtocolHelpers.findRequestMatchingFormIndex(property.forms, this.scheme, this.brokerURI, contentType),
            };
            const formContentType = property.forms[options.formIndex].contentType ?? core_1.ContentSerdes.DEFAULT;
            const inputContent = new core_1.Content(formContentType, stream_1.Readable.from(payload));
            try {
                thing.handleWriteProperty(segments[this.INTERACTION_NAME_SEGMENT_INDEX], inputContent, options);
            }
            catch (err) {
                error(`MqttBrokerServer at ${this.brokerURI} got error on writing to property '${segments[this.INTERACTION_NAME_SEGMENT_INDEX]}': ${err}`);
            }
        }
        else {
            warn(`MqttBrokerServer at ${this.brokerURI} received message for readOnly property at '${segments.join("/")}'`);
        }
    }
    async destroy(thingId) {
        debug(`MqttBrokerServer on port ${this.getPort()} destroying thingId '${thingId}'`);
        let removedThing;
        for (const name of Array.from(this.things.keys())) {
            const expThing = this.things.get(name);
            if (expThing != null && expThing.id != null && expThing.id === thingId) {
                this.things.delete(name);
                removedThing = expThing;
            }
        }
        if (removedThing != null) {
            info(`MqttBrokerServer successfully destroyed '${removedThing.title}'`);
        }
        else {
            info(`MqttBrokerServer failed to destroy thing with thingId '${thingId}'`);
        }
        return removedThing !== undefined;
    }
    async start(servient) {
        if (this.brokerURI === undefined) {
            warn(`No broker defined for MQTT server binding - skipping`);
        }
        else {
            const selfHost = this.config.selfHost ?? false;
            if (selfHost) {
                await this.startBroker();
            }
            if (this.config.psw === undefined) {
                debug(`MqttBrokerServer trying to connect to broker at ${this.brokerURI}`);
            }
            else if (this.config.clientId === undefined) {
                debug(`MqttBrokerServer trying to connect to secured broker at ${this.brokerURI}`);
            }
            else if (this.config.protocolVersion === undefined) {
                debug(`MqttBrokerServer trying to connect to secured broker at ${this.brokerURI} with client ID ${this.config.clientId}`);
            }
            else {
                debug(`MqttBrokerServer trying to connect to secured broker at ${this.brokerURI} with client ID ${this.config.clientId}`);
            }
            try {
                this.broker = await mqtt.connectAsync(this.brokerURI, this.config);
                info(`MqttBrokerServer connected to broker at ${this.brokerURI}`);
                const parsed = new url.URL(this.brokerURI);
                this.address = parsed.hostname;
                const port = parseInt(parsed.port);
                this.port = port > 0 ? port : 1883;
            }
            catch (err) {
                error(`MqttBrokerServer could not connect to broker at ${this.brokerURI}`);
                throw err;
            }
        }
    }
    async stop() {
        if (this.broker !== undefined) {
            this.broker.unsubscribe("*");
            this.broker.end(true);
        }
        if (this.hostedBroker !== undefined) {
            await new Promise((resolve) => this.hostedServer.close(() => resolve()));
            await new Promise((resolve) => this.hostedBroker.close(() => resolve()));
        }
    }
    getPort() {
        return this.port;
    }
    getAddress() {
        return this.address;
    }
    selfHostAuthentication(_client, username, password, done) {
        if (this.config.selfHostAuthentication && username !== undefined) {
            for (let i = 0; i < this.config.selfHostAuthentication.length; i++) {
                if (username === this.config.selfHostAuthentication[i].username &&
                    password.equals(Buffer.from(this.config.selfHostAuthentication[i].password ?? ""))) {
                    done(null, true);
                    return;
                }
            }
            done(null, false);
            return;
        }
        done(null, true);
    }
    async startBroker() {
        return new Promise((resolve, reject) => {
            this.hostedServer = (0, aedes_1.Server)({});
            let server;
            if (this.config.key) {
                server = tls.createServer({ key: this.config.key, cert: this.config.cert }, this.hostedServer.handle);
            }
            else {
                server = net.createServer(this.hostedServer.handle);
            }
            const parsed = new url.URL(this.brokerURI);
            const port = parseInt(parsed.port);
            this.port = port > 0 ? port : 1883;
            this.hostedServer.authenticate = this.selfHostAuthentication.bind(this);
            const errorListener = (err) => {
                error(`error listening for ${this.brokerURI}, ${err}`);
                reject(err);
            };
            server.once("error", errorListener);
            debug(`MqttBrokerServer creating server for ${this.brokerURI}`);
            this.hostedBroker = server.listen(port, parsed.hostname, () => {
                debug(`MqttBrokerServer listening ${this.brokerURI}`);
                server.removeListener("error", errorListener);
                resolve();
            });
        });
    }
}
exports.default = MqttBrokerServer;
